### PR TITLE
Add QuadrupedalRobots

### PR DESCRIPTION
**Pull request by HaojieSHI98**

### CODE REVIEW

**Comment by zenghsh3** [comment]
@HaojieSHI98

**Comment by HaojieSHI98** [comment]
@zenghsh3 I have modified it, please check again, thx.

**Comment by zenghsh3** [comment]
@HaojieSHI98 
- Please remove the `__pycache__` directory.
- Please remove the npy/pt files if they are not used.
- It looks like we don't provide the code of the sim2real part.
- For README.md, we'd better provide a tutorial to introduce how to train a deployable model from scratch.

File: QuadrupedalRobots/ETGRL/alg/BC.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+import parl
+import torch
+from torch.distributions import Normal
+import torch.nn.functional as F
+from copy import deepcopy
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+# __all__ = ['SAC']
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/alg/BC.py:23]: Pls remove these lines.

File: QuadrupedalRobots/ETGRL/alg/BC.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        self.model = model.to(device)
+        self.actor_optimizer = torch.optim.Adam(
+            self.model.get_actor_params(), lr=actor_lr)
+        self.critic_optimizer = torch.optim.Adam(
+            self.model.get_critic_params(), lr=critic_lr)
+
+    def predict(self, obs):
+        act_mean, _ = self.model.policy(obs)
+        action = torch.tanh(act_mean)
+        # print("action:",action)
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/alg/BC.py:48]: Pls remove these lines. +1

File: QuadrupedalRobots/ETGRL/alg/es.py [hunk 3]
```diff
@@ -0,0 +1,614 @@
+import numpy as np
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/alg/es.py:1]: Pls add a reference source.

File: QuadrupedalRobots/ETGRL/alg/sac.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/alg/sac.py:1]: Can we directly import the SAC algorithm from parl?

File: QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        obs,info = self.get_observation()
+        return obs,info
+    
+    def step(self,action,**kwargs):
+        self.robot.Step(action, robot_config.MotorControlMode.POSITION)
+        self.last_action = action
+        self.iter += 1
+        # obs,info = self.get_observation()
+        # info["real_action"] = action
+        # return obs,info
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py:125]: Does the above code work as expected? It looks strange that the step function doesn't return the obs.

File: QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        self._action_filter.reset()
+
+    def _FilterAction(self, action):
+        # initialize the filter history, since resetting the filter will fill
+        # the history with zeros and this can cause sudden movements at the start
+        # of each episode
+        if self._step_counter == 0:
+            default_action = np.array([0,0.9,-1.8]*4)
+            self._action_filter.init_history(default_action)
+            # for j in range(10):
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py:321]: Pls remove these lines.

File: QuadrupedalRobots/ETGRL/deployment/utilities/Bezier.py [hunk 7]
```diff
@@ -0,0 +1,607 @@
+import numpy as np
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/utilities/Bezier.py:1]: Pls add a reference of the third party code.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: QuadrupedalRobots/ETGRL/alg/BC.py (hunk 1 above)
Comment (HaojieSHI98) [reply QuadrupedalRobots/ETGRL/alg/BC.py:23]: ok

File: QuadrupedalRobots/ETGRL/alg/BC.py (hunk 2 above)
Comment (HaojieSHI98) [reply QuadrupedalRobots/ETGRL/alg/BC.py:48]: this function is used.

File: QuadrupedalRobots/ETGRL/alg/sac.py (hunk 4 above)
Comment (HaojieSHI98) [reply QuadrupedalRobots/ETGRL/alg/sac.py:1]: Can't, it‘s a modified version.

File: QuadrupedalRobots/ETGRL/alg/BC.py (hunk 2 above)
Comment (zenghsh3) [reply QuadrupedalRobots/ETGRL/alg/BC.py:48]: just remove the commented lines:
```
# print("action:",action)
```

File: QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py (hunk 5 above)
Comment (HaojieSHI98) [reply QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py:125]: yes, it worked as expected. It's modified and the obs is attained from self.getobs() to save time.

File: QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py (hunk 6 above)
Comment (HaojieSHI98) [reply QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py:321]: > @HaojieSHI98
> 
> * Please remove the `__pycache__` directory.
> * Please remove the npy/pt files if they are not used.
> * It looks like we don't provide the code of the sim2real part.
> * For README.md, we'd better provide a tutorial to introduce how to train a deployable model from scratch.

For the npy/pt files, it's for demonstrate. Shall we remove them?

File: QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+            #     self._action_filter.filter(default_action)
+
+        filtered_action = self._action_filter.filter(action)
+        # print(filtered_action)
+        return filtered_action
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/envs/EnvWrapper.py:326]: Pls remove the following blank lines.

File: QuadrupedalRobots/ETGRL/deployment/robots/minitaur.py [hunk 9]
```diff
@@ -0,0 +1,1435 @@
+# coding=utf-8
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/robots/minitaur.py:1]: Is minitaur*.py needed? Are this codes for minitaur robots?

File: QuadrupedalRobots/ETGRL/deployment/test.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    p = bullet_client.BulletClient(connection_mode=pybullet.DIRECT)
+    p.setAdditionalSearchPath(pybullet_data.getDataPath())
+
+    gait_action = np.load(args.ETG_path)
+    robot = a1_robot.A1Robot(pybullet_client=p, action_repeat=1)
+    env = EnvWrapper.EnvWrapper(robot=robot,dt=args.dt,sensor_mode=sensor_mode,gait=args.gait,
+                                normal= args.normal,enable_action_filter=args.enable_action_filter,
+                                ETG_data = copy(gait_action))
+    obs,info = env.reset()
+    # action = agent.predict(obs)*act_bound
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/test.py:98]: Pls remove these commented codes.

File: QuadrupedalRobots/ETGRL/deployment/test.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    
+    # for i in range(100):
+    #     env.step(np.array([0,0.9,-1.8]*4))
+    #     time.sleep(0.02)
+    # print("ok!")
+    base_act = np.array([0,0.9,-1.8]*4)
+    for i in range(int(args.max_time*100)):
+        t_start = time.clock()
+        ref_action = gait_action[i]
+        # if args.sensor_ETG:
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/deployment/test.py:108]: Pls remove these commented codes.

File: QuadrupedalRobots/ETGRL/model/Dynamic_parallel_model.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+            mean_re = self.update(epoch) 
+            if epoch % 5 ==0 and epoch > 0:
+                r = self.evaluate_episode(epoch)
+                self.save(epoch,self.outdir)
+    def evaluate_episode(self,epoch):
+        results = self.agent_list[0].sample_episode(param=self.dynamic_param,key="height")
+        reward,_ = results.get()
+        logger.info('Steps: {} Eval Reward: {}'.format(epoch, reward))
+        summary.add_scalar('eval/episode_reward', reward, epoch)
+        return reward
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/model/Dynamic_parallel_model.py:196]: Pls remove the following blank lines.

File: QuadrupedalRobots/ETGRL/README.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+| ES               | bool     | Use dual training procedure or not.         |
+| es_rpm               | bool     | Store experience in replay buffer during ES phase or not.        |
+| eval               | bool     | Evaluation or training.       |
+| render               | bool     | Render or not.        |
+| act_bound               | float     | Action bound for neural network policy.        |
+| load               | string     | The path of neural network model to load.        |
+
+For example, train ETG-RL in stairstair.
+```python
+python train.py --task_mode stairstair --ETG_path data/origin_ETG/ESStair_origin.npz
```
Comment (zenghsh3) [inline QuadrupedalRobots/ETGRL/README.md:77]: Pls specify why we need the `ESStair_origin.npz`, and how to generate it.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: QuadrupedalRobots/ETGRL/deployment/robots/minitaur.py (hunk 9 above)
Comment (HaojieSHI98) [reply QuadrupedalRobots/ETGRL/deployment/robots/minitaur.py:1]: it's needed, the motion imitation modify the code from minitaur, it's parent class.

**Review by zenghsh3 — APPROVED** [review]
LGTM

**Pull request merged**